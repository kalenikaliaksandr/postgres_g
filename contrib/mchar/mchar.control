# mchar extension
comment = 'SQL Server text type'
default_version = '2.0'
module_pathname = '$libdir/mchar'
relocatable = true

