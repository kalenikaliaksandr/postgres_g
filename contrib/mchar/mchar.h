#ifndef __MCHAR_H__
#define __MCHAR_H__

#include "postgres.h"
#include "mb/pg_wchar.h"
#include "utils/builtins.h"
#include "unicode/uchar.h"
#include "unicode/ustring.h"

typedef struct {
	int32	len;
	int32 	typmod;
	UChar	data[1];
} MChar;

#define MCHARHDRSZ	offsetof(MChar, data)	
#define MCHARLENGTH(m)	( VARSIZE(m)-MCHARHDRSZ )
#define UCHARLENGTH(m)  ( MCHARLENGTH(m)/sizeof(UChar) )	

#define DatumGetMChar(m)	((MChar*)DatumGetPointer(m))
#define MCharGetDatum(m)	PointerGetDatum(m)

#define	PG_GETARG_MCHAR(n)	DatumGetMChar(PG_DETOAST_DATUM(PG_GETARG_DATUM(n)))
#define	PG_GETARG_MCHAR_COPY(n)	DatumGetMChar(PG_DETOAST_DATUM_COPY(PG_GETARG_DATUM(n)))

#define PG_RETURN_MCHAR(m)	PG_RETURN_POINTER(m)

typedef struct {
	int32	len;
	UChar	data[1];
} MVarChar;

#define MVARCHARHDRSZ 	offsetof(MVarChar, data)
#define MVARCHARLENGTH(m)  ( VARSIZE(m)-MVARCHARHDRSZ )
#define UVARCHARLENGTH(m)  ( MVARCHARLENGTH(m)/sizeof(UChar) )

#define DatumGetMVarChar(m)	((MVarChar*)DatumGetPointer(m))
#define MVarCharGetDatum(m)	PointerGetDatum(m)

#define	PG_GETARG_MVARCHAR(n)	DatumGetMVarChar(PG_DETOAST_DATUM(PG_GETARG_DATUM(n)))
#define	PG_GETARG_MVARCHAR_COPY(n)	DatumGetMVarChar(PG_DETOAST_DATUM_COPY(PG_GETARG_DATUM(n)))

#define PG_RETURN_MVARCHAR(m)	PG_RETURN_POINTER(m)


int Char2UChar(const char * src, int srclen, UChar *dst); 
int UChar2Char(const UChar * src, int srclen, char *dst); 
int UChar2Wchar(UChar * src, int srclen, pg_wchar *dst);
int UCharCompare(UChar * a, int alen, UChar *b, int blen); 
int UCharCaseCompare(UChar * a, int alen, UChar *b, int blen); 

void FillWhiteSpace( UChar *dst, int n );

int lengthWithoutSpaceVarChar(MVarChar *m);
int lengthWithoutSpaceChar(MChar *m);

extern Datum       mchar_hash(PG_FUNCTION_ARGS);
extern Datum       mvarchar_hash(PG_FUNCTION_ARGS);

int m_isspace(UChar c); /* is == ' ' */

Datum hash_uchar( UChar *s, int len );
#endif
