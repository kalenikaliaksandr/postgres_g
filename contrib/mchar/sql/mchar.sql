-- I/O tests

select '1'::mchar;
select '2  '::mchar;
select '10          '::mchar;

select '1'::mchar(2);
select '2 '::mchar(2);
select '3  '::mchar(2);
select '10          '::mchar(2);

select '                  '::mchar(10); 
select '                  '::mchar; 

-- operations & functions

select length('1'::mchar);
select length('2  '::mchar);
select length('10          '::mchar);

select length('1'::mchar(2));
select length('2 '::mchar(2));
select length('3  '::mchar(2));
select length('10          '::mchar(2));

select length('                  '::mchar(10)); 
select length('                  '::mchar); 

select 'asd'::mchar(10) || '>'::mchar(10);
select length('asd'::mchar(10) || '>'::mchar(10));
select 'asd'::mchar(2)  || '>'::mchar(10);
select length('asd'::mchar(2) || '>'::mchar(10));

-- Comparisons

select 'asdf'::mchar = 'aSdf'::mchar;
select 'asdf'::mchar = 'aSdf '::mchar;
select 'asdf'::mchar = 'aSdf 1'::mchar(4);
select 'asdf'::mchar = 'aSdf 1'::mchar(5);
select 'asdf'::mchar = 'aSdf 1'::mchar(6);
select 'asdf'::mchar(3) = 'aSdf 1'::mchar(5);
select 'asdf'::mchar(3) = 'aSdf 1'::mchar(3);

select 'asdf'::mchar < 'aSdf'::mchar;
select 'asdf'::mchar < 'aSdf '::mchar;
select 'asdf'::mchar < 'aSdf 1'::mchar(4);
select 'asdf'::mchar < 'aSdf 1'::mchar(5);
select 'asdf'::mchar < 'aSdf 1'::mchar(6);

select 'asdf'::mchar <= 'aSdf'::mchar;
select 'asdf'::mchar <= 'aSdf '::mchar;
select 'asdf'::mchar <= 'aSdf 1'::mchar(4);
select 'asdf'::mchar <= 'aSdf 1'::mchar(5);
select 'asdf'::mchar <= 'aSdf 1'::mchar(6);

select 'asdf'::mchar >= 'aSdf'::mchar;
select 'asdf'::mchar >= 'aSdf '::mchar;
select 'asdf'::mchar >= 'aSdf 1'::mchar(4);
select 'asdf'::mchar >= 'aSdf 1'::mchar(5);
select 'asdf'::mchar >= 'aSdf 1'::mchar(6);

select 'asdf'::mchar > 'aSdf'::mchar;
select 'asdf'::mchar > 'aSdf '::mchar;
select 'asdf'::mchar > 'aSdf 1'::mchar(4);
select 'asdf'::mchar > 'aSdf 1'::mchar(5);
select 'asdf'::mchar > 'aSdf 1'::mchar(6);

select max(ch) from chvch;
select min(ch) from chvch;

select substr('1234567890'::mchar, 3) = '34567890' as "34567890";
select substr('1234567890'::mchar, 4, 3) = '456' as "456";

select lower('asdfASDF'::mchar);
select upper('asdfASDF'::mchar);

select 'asd'::mchar == 'aSd'::mchar;
select 'asd'::mchar == 'aCd'::mchar;
select 'asd'::mchar == NULL;
select NULL == 'aCd'::mchar;
select NULL::mchar == NULL;


--Note: here we use different space symbols, be carefull to copy it!
select v, count(*) from
(values  (1, '4 242'::mchar), (2, '4 242'), (3, 'aSDF'), (4, 'asdf')) as t(i,v) group by v;
set enable_hashagg=off;
select v, count(*) from
(values  (1, '4 242'::mchar), (2, '4 242'), (3, 'aSDF'), (4, 'asdf')) as t(i,v) group by v;
reset enable_hashagg;
