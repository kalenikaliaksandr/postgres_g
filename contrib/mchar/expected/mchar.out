-- I/O tests
select '1'::mchar;
 mchar 
-------
 1
(1 row)

select '2  '::mchar;
 mchar 
-------
 2
(1 row)

select '10          '::mchar;
 mchar 
-------
 10
(1 row)

select '1'::mchar(2);
 mchar 
-------
 1 
(1 row)

select '2 '::mchar(2);
 mchar 
-------
 2 
(1 row)

select '3  '::mchar(2);
 mchar 
-------
 3 
(1 row)

select '10          '::mchar(2);
 mchar 
-------
 10
(1 row)

select '                  '::mchar(10); 
   mchar    
------------
           
(1 row)

select '                  '::mchar; 
 mchar 
-------
 
(1 row)

-- operations & functions
select length('1'::mchar);
 length 
--------
      1
(1 row)

select length('2  '::mchar);
 length 
--------
      1
(1 row)

select length('10          '::mchar);
 length 
--------
      2
(1 row)

select length('1'::mchar(2));
 length 
--------
      1
(1 row)

select length('2 '::mchar(2));
 length 
--------
      1
(1 row)

select length('3  '::mchar(2));
 length 
--------
      1
(1 row)

select length('10          '::mchar(2));
 length 
--------
      2
(1 row)

select length('                  '::mchar(10)); 
 length 
--------
      0
(1 row)

select length('                  '::mchar); 
 length 
--------
      0
(1 row)

select 'asd'::mchar(10) || '>'::mchar(10);
       ?column?       
----------------------
 asd       >         
(1 row)

select length('asd'::mchar(10) || '>'::mchar(10));
 length 
--------
     11
(1 row)

select 'asd'::mchar(2)  || '>'::mchar(10);
   ?column?   
--------------
 as>         
(1 row)

select length('asd'::mchar(2) || '>'::mchar(10));
 length 
--------
      3
(1 row)

-- Comparisons
select 'asdf'::mchar = 'aSdf'::mchar;
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar = 'aSdf '::mchar;
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar = 'aSdf 1'::mchar(4);
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar = 'aSdf 1'::mchar(5);
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar = 'aSdf 1'::mchar(6);
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar(3) = 'aSdf 1'::mchar(5);
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar(3) = 'aSdf 1'::mchar(3);
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar < 'aSdf'::mchar;
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar < 'aSdf '::mchar;
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar < 'aSdf 1'::mchar(4);
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar < 'aSdf 1'::mchar(5);
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar < 'aSdf 1'::mchar(6);
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar <= 'aSdf'::mchar;
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar <= 'aSdf '::mchar;
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar <= 'aSdf 1'::mchar(4);
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar <= 'aSdf 1'::mchar(5);
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar <= 'aSdf 1'::mchar(6);
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar >= 'aSdf'::mchar;
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar >= 'aSdf '::mchar;
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar >= 'aSdf 1'::mchar(4);
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar >= 'aSdf 1'::mchar(5);
 ?column? 
----------
 t
(1 row)

select 'asdf'::mchar >= 'aSdf 1'::mchar(6);
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar > 'aSdf'::mchar;
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar > 'aSdf '::mchar;
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar > 'aSdf 1'::mchar(4);
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar > 'aSdf 1'::mchar(5);
 ?column? 
----------
 f
(1 row)

select 'asdf'::mchar > 'aSdf 1'::mchar(6);
 ?column? 
----------
 f
(1 row)

select max(ch) from chvch;
     max      
--------------
 One space   
(1 row)

select min(ch) from chvch;
     min      
--------------
 1 space     
(1 row)

select substr('1234567890'::mchar, 3) = '34567890' as "34567890";
 34567890 
----------
 f
(1 row)

select substr('1234567890'::mchar, 4, 3) = '456' as "456";
 456 
-----
 t
(1 row)

select lower('asdfASDF'::mchar);
  lower   
----------
 asdfasdf
(1 row)

select upper('asdfASDF'::mchar);
  upper   
----------
 ASDFASDF
(1 row)

select 'asd'::mchar == 'aSd'::mchar;
 ?column? 
----------
 t
(1 row)

select 'asd'::mchar == 'aCd'::mchar;
 ?column? 
----------
 f
(1 row)

select 'asd'::mchar == NULL;
 ?column? 
----------
 f
(1 row)

select NULL == 'aCd'::mchar;
 ?column? 
----------
 f
(1 row)

select NULL::mchar == NULL;
 ?column? 
----------
 t
(1 row)

--Note: here we use different space symbols, be carefull to copy it!
select v, count(*) from
(values  (1, '4 242'::mchar), (2, '4 242'), (3, 'aSDF'), (4, 'asdf')) as t(i,v) group by v;
   v   | count 
-------+-------
 aSDF  |     2
 4 242 |     2
(2 rows)

set enable_hashagg=off;
select v, count(*) from
(values  (1, '4 242'::mchar), (2, '4 242'), (3, 'aSDF'), (4, 'asdf')) as t(i,v) group by v;
   v   | count 
-------+-------
 4 242 |     2
 aSDF  |     2
(2 rows)

reset enable_hashagg;
